### PR TITLE
Future WIP

### PR DESCRIPTION
**Pull request by cabelitos**

### CODE REVIEW

File: src/lib/ecore/efl_promise2.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+extern "C" {
+#endif
+
+#include <Eina.h>
+
+typedef struct _Efl_Future2_Desc Efl_Future2_Desc;
+typedef struct _Efl_Promise2 Efl_Promise2;
+typedef struct _Efl_Future2 Efl_Future2;
+typedef Eina_Value *(*Efl_Future2_Cb)(void *data, const Eina_Value *value);
+typedef void (*Efl_Future2_Progress_Cb)(void *data);
```
Comment (barbieri) [inline src/lib/ecore/efl_promise2.h]: you can remove the progress one, it's gone

File: src/lib/ecore/efl_promise2.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#ifndef _EFL_PROMISE2_H_
+#define _EFL_PROMISE2_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <Eina.h>
+
+typedef struct _Efl_Future2_Desc Efl_Future2_Desc;
```
Comment (barbieri) [inline src/lib/ecore/efl_promise2.h]: document these and fields, as well as the functions and parameters so we don't forget :-)

File: src/examples/ecore/ecore_promise2_example.c [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+ err_promise:
+   free(ctx);
+   return NULL;
+}
+
+static Eina_Value *
+_simple_ok(void *data EINA_UNUSED, const Eina_Value *v)
+{
+   const char *msg;
+   eina_value_get(v, &msg);
```
Comment (barbieri) [inline src/examples/ecore/ecore_promise2_example.c]: prior to every value access, check if the value is of the expected type:

```c
if (eina_value_type_get(v) == EINA_VALUE_TYPE_STRING)
  {
      const char *msg;
      eina_value_get(v, &msg);
      ...
  }
```
We may add some helpers to https://git.enlightenment.org/core/efl.git/tree/src/lib/eina/eina_inline_value_util.x such as:

```c
if (eina_value_util_string_get(v, &msg)) // check & get
   {
...
   }
```

File: src/examples/ecore/ecore_promise2_example.c [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+{
+   Eina_Bool *should_fail = data;
+   Eina_Value *new_v = NULL;
+
+   if (*should_fail)
+     {
+        *should_fail = EINA_FALSE;
+        new_v = eina_value_new(EINA_VALUE_TYPE_ERROR);
+        EINA_SAFETY_ON_NULL_RETURN_VAL(new_v, NULL);
+        eina_value_set(new_v, ENETDOWN);
```
Comment (barbieri) [inline src/examples/ecore/ecore_promise2_example.c]: add a helper to https://git.enlightenment.org/core/efl.git/tree/src/lib/eina/eina_inline_value_util.x so this becomes an one-liner.

**Comment by barbieri** [comment]
talking to Cedric we may drop `error_cb` and `free_cb`, handling all in a single callback. This way we save some code and memory (2 pointers) as well as remove the problems related to calling back the user 2 times (and the nasties that he can do).

that said, you can also simplify the `efl_future2_then()` without the `Efl_Future2_Desc`:

```c
typedef Eina_Value * (*Efl_Future2_Cb)(const void *data, const Eina_Value *value, const Efl_Future2 *dead_future_ptr);

Efl_Future2 *efl_future2_then(Efl_Future2 *f, Efl_Future2_Cb then_cb, const void *data);
```

File: src/lib/ecore/efl_promise2.c [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+   if (value && value->type == EINA_VALUE_TYPE_ERROR)
+     rejected = EINA_TRUE;
+
+   while (f)
+     {
+        if (((!rejected && f->success) || (rejected && f->error)))
+          break;
+        Efl_Future2 *n = f->next;
+        _efl_future2_free(f);
+        f = n;
```
Comment (cabelitos) [inline src/lib/ecore/efl_promise2.c]: Indeed

**Pull request closed without merging**